algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+import algebra.ring algebra.field data.set.basic order.zorn tactic.norm_num tactic.ring
+
+universes u v w
+
+-- page 1
+
+class zero_ring (α : Type u) [comm_ring α] : Prop :=
+(eq_zero : ∀ x:α, x = 0)
+
+instance zero_ring_of_zero_eq_one {α : Type u} [comm_ring α] : (0:α)=1 → zero_ring α
+| h := ⟨λ x, calc
+  x = x * 1   : by simp
+  ... = x * 0 : by rw h
+  ... = 0     : by simp⟩
+
+
+-- page 2
+
+class is_hom {α : Type u} {β : Type v} [comm_ring α] [comm_ring β] (f : α → β) : Prop :=
+(map_add : ∀ {x y}, f (x + y) = f x + f y)
+(map_mul : ∀ {x y}, f (x * y) = f x * f y)
+(map_one : f 1 = 1)
+
+namespace is_hom
+
+variables {α : Type u} {β : Type v} [comm_ring α] [comm_ring β]
+variables (f : α → β) [is_hom f] {x y : α}
+
+attribute [simp] map_add
+attribute [simp] map_mul
+attribute [simp] map_one
+
+@[simp] lemma map_zero : f 0 = 0 :=
+calc f 0 = f (0 + 0) - f 0 : by rw [map_add f]; simp
+     ... = 0 : by simp
+
+@[simp] lemma map_neg : f (-x) = -f x :=
+calc f (-x) = f (-x + x) - f x : by rw [map_add f]; simp
+        ... = -f x : by simp [map_zero f]
+
+@[simp] lemma map_sub : f (x - y) = f x - f y :=
+by simp [map_add f, map_neg f]
+
+end is_hom
+
+class subring (α : Type u) [comm_ring α] (S : set α) : Prop :=
+(add_mem : ∀ {x y}, x ∈ S → y ∈ S → x + y ∈ S)
+(neg_mem : ∀ {x}, x ∈ S → -x ∈ S)
+(mul_mem : ∀ {x y}, x ∈ S → y ∈ S → x * y ∈ S)
+(one_mem : (1:α) ∈ S)
+
+namespace subring
+
+variables (α : Type u) [comm_ring α] (S : set α) [subring α S]
+
+instance : comm_ring S :=
+{ add            := λ ⟨x, hx⟩ ⟨y, hy⟩, ⟨x + y, add_mem hx hy⟩,
+  add_assoc      := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ add_assoc x y z,
+  zero           := ⟨0, eq.subst (add_neg_self (1:α)) $ add_mem (one_mem S) $ neg_mem $ one_mem S⟩,
+  zero_add       := λ ⟨x, hx⟩, subtype.eq $ zero_add x,
+  add_zero       := λ ⟨x, hx⟩, subtype.eq $ add_zero x,
+  neg            := λ ⟨x, hx⟩, ⟨-x, neg_mem hx⟩,
+  add_left_neg   := λ ⟨x, hx⟩, subtype.eq $ add_left_neg x,
+  add_comm       := λ ⟨x, hx⟩ ⟨y, hy⟩, subtype.eq $ add_comm x y,
+  mul            := λ ⟨x, hx⟩ ⟨y, hy⟩, ⟨x * y, mul_mem hx hy⟩,
+  mul_assoc      := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ mul_assoc x y z,
+  one            := ⟨1, one_mem S⟩,
+  one_mul        := λ ⟨x, hx⟩, subtype.eq $ one_mul x,
+  mul_one        := λ ⟨x, hx⟩, subtype.eq $ mul_one x,
+  left_distrib   := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ left_distrib x y z,
+  right_distrib  := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ right_distrib x y z,
+  mul_comm       := λ ⟨x, hx⟩ ⟨y, hy⟩, subtype.eq $ mul_comm x y }
+
+@[simp] lemma add (x y : α) (hx : x ∈ S) (hy : y ∈ S) :
+(⟨x, hx⟩ : S) + ⟨y, hy⟩ = ⟨x + y, add_mem hx hy⟩ := rfl
+
+@[simp] lemma mul (x y : α) (hx : x ∈ S) (hy : y ∈ S) :
+(⟨x, hx⟩ : S) * ⟨y, hy⟩ = ⟨x * y, mul_mem hx hy⟩ := rfl
+
+instance is_hom : is_hom ((λ x, x) : S → α) :=
```

Comment:
Rename `is_hom`, I suggest to call the instance `is(_ring)_hom_subtype_val` and also use `subtype.val` in the type.
For two reasons:
1) It hides what is really going on
2) We might also w want the `subtype.mk` as hom, i.e. each hom `f : β → α`, with range in `S` induces a `f' : β → S`